### PR TITLE
App permissions with action patterns do not retrieve privileges

### PR DESCRIPTION
**Pull request by albertzaharovits**
In order for file realm users, with application privileges, to continue
to work (authenticate) when the `.security` index is unavailable,
Security must avoid loading the application privileges from the .security index
(which is unavailable, hence error out).
Roles (defined inside the `roles.yml` file, not in the index) that must work
when the `.security` index is unavailable must use inlined action patterns
for application permisions, rather than privilege names.

Related https://github.com/elastic/elasticsearch/issues/85312

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java [hunk 1]
```diff
@@ -162,10 +162,17 @@ public static void validatePrivilegeName(String name) {
         }
     }
 
-    private static boolean isValidPrivilegeName(String name) {
+    public static boolean isValidPrivilegeName(String name) {
         return VALID_NAME.matcher(name).matches();
     }
 
+    public static void validateActionName(String action) {
```
Comment (justincr-elastic) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java:169]: Does it make sense for a new `validateActionName` method to call a new `isValidActionName` method?

For example, there is an existing `validatePrivilegeName` method which calls `isValidPrivilegeName`. Does it make sense to follow that pattern?

In other words, I found it confusing to have a method called `validateActionName` which does validation on action name, but then also calls through to `validatePrivilegeOrActionName`.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesRequest.java [hunk 2]
```diff
@@ -63,14 +63,8 @@ public ActionRequestValidationException validate() {
...
                 for (String action : privilege.getActions()) {
-                    if (action.indexOf('/') == -1 && action.indexOf('*') == -1 && action.indexOf(':') == -1) {
-                        validationException = addValidationError(
-                            "action [" + action + "] must contain one of [ '/' , '*' , ':' ]",
-                            validationException
-                        );
-                    }
                     try {
-                        ApplicationPrivilege.validatePrivilegeOrActionName(action);
+                        ApplicationPrivilege.validateActionName(action);
```
Comment (justincr-elastic) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesRequest.java:67]: I found it confusing to have a method called `validateActionName` which does validation on action name, but then also calls through to `validatePrivilegeOrActionName`.

**Review by justincr-elastic — CHANGES_REQUESTED** [review]
I would need more information to be able to complete a review, but I did have a feedback point.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java (hunk 1 above)
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java:169]: > Does it make sense for a new validateActionName method to call a new isValidActionName method?

I am unsure what that would achieve.
The `validateActionName` throws two different kinds of errors at the moment. How can the proposal, as I understand it, achieve the same thing when just calling `isValidPrivilegeName`?

I have pushed dea850d34b7d04d1e338001 . Not sure if this helps or not.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesRequest.java (hunk 2 above)
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/privilege/PutPrivilegesRequest.java:67]: Do you also have proposals to improve it?
Also keeping in mind this is a bug-fix PR.
I only slightly changed code in this part to make sure the action names I generate IN TESTS are valid.

**Comment by albertzaharovits** [comment]
This PR fixes the bug from https://github.com/elastic/elasticsearch/issues/85312 by making any file-based role that contains application permissions that do not refer to named application privileges (which only mention inlined action patterns, like the `superuser` role does) not unnecessarily query the `.security` index. By avoiding querying the index, such roles can be used even when the index is unavailable.

Another option to fix the same bug would be to always return the statically built `superuser` role, after failing to resolve a role name list that includes it. We do something like that today, when failing to retrieve any role descriptors, but not when failing to retrieve privileges referred in those role descriptors.

I think both approaches have pros-cons, but I don't think we have to choose between them.
I think it is perfectly reasonable for file-based roles to not reach the index if they don't need to, but if they do, and fail, we should not fail the authentication if the user also has `superuser`.

**Comment by tvernum** [comment]
Hey @albertzaharovits - this approach looks like a good direction to me. 
Are you still looking for a review?

**Review by justincr-elastic — APPROVED** [review]
LGTM

**Comment by albertzaharovits** [comment]
> Hey @albertzaharovits - this approach looks like a good direction to me.
Are you still looking for a review?

Yes, please review. I have adapted the PR description and label to reflect that this is now `>enhancement` rather than a `>bug` fix after you have merged https://github.com/elastic/elasticsearch/pull/85519 .

**Comment by tvernum** [comment]
Given this is now an enhancement, I don't think we should backport to `7.17`

**Review by tvernum — APPROVED** [review]
LGTM

**Comment by albertzaharovits** [comment]
@elasticmachine update branch

**Pull request merged**